File: src/converters.cpp
#include "converters.hpp"
#include "utils/logging.hpp"

extern logging::Logger logger;

namespace converters {
namespace {

std::string getId(json node) {
    return std::to_string(node["id"].get<int64_t>());
}

Window toWindow(json const &node) {
    json const &rect = node["rect"];
    bool floating = node["type"] == "floating_con";
    bool focused = floating ? node["nodes"][0]["focused"] : node["focused"];
    std::string id = floating ? getId(node["nodes"][0]) : getId(node);
    return {id, rect["x"], rect["y"], rect["width"], rect["height"], focused, floating};
}

planar::Window toPlanar(Window const &window) {
    return {{window.x, window.y, window.w, window.h}, window.id};
}

} // namespace

json focused_subnode(json &node) {
    auto &focus = node["focus"];
    if (focus == nullptr or focus.empty()) {
        return nullptr;
    }
    int64_t focus_id = focus[0];
    auto &nodes = node["nodes"];
    auto it = std::find_if(nodes.begin(), nodes.end(), [focus_id](json const &node) { return node["id"] == focus_id; });
    if (it == nodes.end()) {
        return nullptr;
    }
    return *it;
}

std::vector<json> visible_nodes(json node) {
    logger.debug("Node iterated id:%ld type:%s layout:%s",
                 node["id"].get<int64_t>(),
                 node["type"].get<std::string>().c_str(),
                 node["layout"].get<std::string>().c_str());
    if (node["nodes"].empty() and node["type"] == "con") {
        if (node["rect"]["width"] == 0 or node["rect"]["height"] == 0) {
            return {};
        }
        return {node};
    }
    std::string layout = node["layout"];
    if (layout == "splith" or layout == "splitv" or layout == "output") {
        std::vector<json> result;
        auto &floating_nodes = node["floating_nodes"];
        result.insert(result.end(), floating_nodes.begin(), floating_nodes.end());
        for (auto &subnode : node["nodes"]) {
            auto leaves = visible_nodes(subnode);
            result.insert(result.end(), leaves.begin(), leaves.end());
        }
        return result;
    } else if (layout == "tabbed" or layout == "stacked") {
        auto focused = focused_subnode(node);
        if (focused != nullptr) {
            return visible_nodes(focused);
        }
    } else if (layout == "dockarea") {
        return {};
    } else {
        logger.critical("Unsupported layout:%s found for id: %d",
                        layout.c_str(), getId(node));
    }
    return {};
}

std::vector<Window> to_windows(std::vector<json> const &nodes) {
    std::vector<Window> windows;
    std::transform(nodes.begin(), nodes.end(), std::back_inserter(windows), &toWindow);
    return windows;
}

std::vector<Window> floating_windows(std::vector<Window> const &windows) {
    std::vector<Window> result;
    std::copy_if(windows.begin(), windows.end(), std::back_inserter(result),
                 [](Window const &window) { return window.floating; });
    return result;
}

std::vector<Window> tiled_windows(std::vector<Window> const &windows) {
    std::vector<Window> result;
    std::copy_if(windows.begin(), windows.end(), std::back_inserter(result),
                 [](Window const &window) { return not window.floating; });
    return result;
}

bool any_focused(const std::vector<Window> &windows) {
    auto it = std::find_if(windows.begin(), windows.end(), [](const Window &window){ return window.focused; });
    return it != windows.end();
}

planar::Arrangement visible_grid(const std::vector<Window> &nodes) {
    std::vector<planar::Window> windows;
    std::transform(nodes.begin(), nodes.end(), std::back_inserter(windows),
                   &toPlanar);
    auto it = std::find_if(nodes.begin(), nodes.end(), [](Window const &window) { return window.focused; });
    if (it == nodes.end()) {
        logger.warning("No focused node found out of %lu", nodes.size());
        it = nodes.begin();
    }
    size_t index = std::distance(nodes.begin(), it);
    return planar::Arrangement(std::move(windows), index, planar::Relation::BORDER);
}

json find_deepest_focused_tabbed(json node) {
    logger.debug("Node iterated id:%s type:%s layout:%s",
                 getId(node).c_str(),
                 node["type"].get<std::string>().c_str(),
                 node["layout"].get<std::string>().c_str());
    auto focused = focused_subnode(node);
    if (focused == nullptr) {
        return {};
    }
    auto result = find_deepest_focused_tabbed(focused);
    if (result != nullptr) {
        return result;
    }
    std::string layout = node["layout"];
    if (layout == "tabbed" or layout == "stacked") {
        return node;
    }
    return {};
}

json find_deepest_focused(json node) {
    logger.debug("Node iterated id:%s type:%s layout:%s",
                 getId(node).c_str(),
                 node["type"].get<std::string>().c_str(),
                 node["layout"].get<std::string>().c_str());
    auto focused = focused_subnode(node);
    if (focused == nullptr) {
        return node;
    }
    auto result = find_deepest_focused(focused);
    if (result != nullptr) {
        return result;
    }
    return focused;
}

linear::Sequence available_tabs(json node) {
    node = find_deepest_focused_tabbed(node);
    if (node == nullptr) {
        return {{},0};
    }
    json nodes = node["nodes"];
    if (nodes == nullptr or nodes.empty()) {
        return {{},0};
    }

    std::vector<json> leaves;
    std::transform(nodes.begin(), nodes.end(), std::back_inserter(leaves),
                   find_deepest_focused);

    std::vector<std::string> tabs;
    std::transform(leaves.begin(), leaves.end(), std::back_inserter(tabs), &getId);

    int64_t focus_id = node["focus"][0].get<int64_t>();
    auto it = std::find_if(nodes.begin(), nodes.end(),
                           [focus_id](json const &node) { return node["id"] == focus_id; });
    size_t index = std::distance(nodes.begin(), it);
    return {tabs, index};
}

} // namespace converters

File: rust/src/converters.rs
/// This file provides functions to convert JSON nodes representing windows in a window manager's
/// tree structure. It includes functions to extract visible nodes, available tabs, floating and
/// tiled windows, and to convert these nodes into a structured format for further processing.
/// It also includes functions to determine the layout of nodes and to find focused windows in
/// the tree structure.
///
/// This module is part of a window manager's arrangement system, which allows for
/// manipulating and querying the layout of windows in a graphical user interface.

use crate::linear;
use crate::planar;
use crate::logging::OptionExt;
use crate::types::Windows;

/// Returns a collection of windows that are floating, i.e., those that are not tiled.
pub fn floating(windows: &Windows) -> Windows {
    windows.iter()
        .filter(|w| w.floating)
        .cloned()
        .collect()
}

/// Returns a collection of windows that are not floating, i.e., those that are tiled.
pub fn tiled(windows: &Windows) -> Windows {
    windows.iter()
        .filter(|w| !w.floating)
        .cloned()
        .collect()
}

/// Converts a JSON node to a `planar::Arrangement`.
/// This function assumes that the node represents a workspace or root node
/// and contains a list of windows.
pub fn as_arrangement(windows: Windows, relation: planar::Relation) -> planar::Arrangement {
    let current = focused_index(&windows).unwrap_or(0);
    planar::Arrangement::new(windows, Some(current), Some(relation))
}

/// Converts a collection of `Windows` to a `linear::Sequence`.
/// This function creates a sequence of window IDs and marks the focused window by its index.
pub fn as_sequence(windows: &Windows) -> linear::Sequence {
    let focused = focused_index(windows).unwrap_or(0);
    return linear::Sequence::new(windows.iter().map(|w| w.id).collect(), focused);
}

/// Returns whether any window in the provided `Windows` is focused.
pub fn any_focused(windows: &Windows) -> bool {
    windows.iter().any(|w| w.focused)
}

/// Returns the index of the currently focused window, if any.
fn focused_index(windows: &Windows) -> Option<usize> {
    windows.iter().position(|w| w.focused).wanted(
        format!("No focused window found in windows: {:?}", windows).as_str())
}

#[cfg(test)]
mod tests {
    use crate::types::Rect;
    use crate::types::Window;
    use super::*;

    /// Tests for floating and tiled windows.
    #[test]
    fn test_floating_and_tiled() {
        let windows = vec![
            Window { id: 1, rect: Rect {x: 0, y: 0, w: 100, h: 100}, focused: true, floating: false },
            Window { id: 2, rect: Rect {x: 100, y: 100, w: 200, h: 200}, focused: false, floating: true },
        ];
        let floating_windows = floating(&windows);
        let tiled_windows = tiled(&windows);
        assert_eq!(floating_windows.len(), 1);
        assert_eq!(tiled_windows.len(), 1);
    }

    /// Tests for checking if any window is focused.
    #[test]
    fn test_any_focused() {
        let windows = vec![
            Window { id: 1, rect: Rect {x: 0, y: 0, w: 100, h: 100}, focused: true, floating: false },
            Window { id: 2, rect: Rect {x: 100, y: 100, w: 200, h: 200}, focused: false, floating: true },
        ];
        assert!(any_focused(&windows));
    }

    /// Tests for visible nodes extraction.
    #[test]
    fn test_as_arrangement() {
        let windows = vec![
            Window { id: 1, rect: Rect { x: 0, y: 0, w: 100, h: 100 }, focused: true, floating: false },
            Window { id: 2, rect: Rect { x: 100, y: 100, w: 200, h: 200 }, focused: false, floating: true },
        ];
        let arrangement = as_arrangement(windows, planar::Relation::Border);
        assert_eq!(arrangement.windows.len(), 2);
        assert_eq!(arrangement.current, 0);
    }

    /// Tests for focused index extraction.
    /// We expect the function to return the index of the first focused window in the provided
    #[test]
    fn test_focused_index() {
        let windows = vec![
            Window { id: 1, rect: Rect { x: 0, y: 0, w: 100, h: 100 }, focused: true, floating: false },
            Window { id: 2, rect: Rect { x: 100, y: 100, w: 200, h: 200 }, focused: false, floating: true },
        ];
        assert_eq!(focused_index(&windows), Some(0));

        let windows = vec![
            Window { id: 1, rect: Rect { x: 0, y: 0, w: 100, h: 100 }, focused: false, floating: false },
            Window { id: 2, rect: Rect { x: 100, y: 100, w: 200, h: 200 }, focused: false, floating: true },
        ];
        assert_eq!(focused_index(&windows), None);
    }

    /// Tests sequence conversion.
    /// We expect the function to return a sequence of window IDs and the index of the focused
    /// window.
    #[test]
    fn test_as_sequence() {
        let windows = vec![
            Window { id: 1, rect: Rect { x: 0, y: 0, w: 100, h: 100 }, focused: true, floating: false },
            Window { id: 2, rect: Rect { x: 100, y: 100, w: 200, h: 200 }, focused: false, floating: true },
        ];
        let sequence = as_sequence(&windows);
        assert_eq!(sequence[0], 1);
        assert_eq!(sequence[1], 2);
    }
}
